File: r/R/postprocess.R
#' Post-Process Composite Scores
#'
#' Applies the three-step post-processing pipeline to point estimates and CI
#' bounds: (1) global z-scoring, (2) Box-Cox transformation with
#' \eqn{\lambda = 0.5}, (3) min-max rescaling to \eqn{[0, 100]}.
#'
#' @details
#' **Step 1 – Z-score**: subtract the mean and divide by SD of the point
#' estimates.  The same shift and scale are applied to CI bounds.
#'
#' **Step 2 – Box-Cox** (\eqn{\lambda = 0.5}): scores are shifted so the
#' minimum exceeds zero (\eqn{x \leftarrow x + |\min(x)| + \epsilon}),
#' then \eqn{y = (x^\lambda - 1)/\lambda} is applied.  The same affine
#' shift (from the point estimates) is used for CI bounds before applying
#' the same power transform, so that bounds stay monotonically ordered
#' relative to the point estimate.
#'
#' **Step 3 – Min-max**: \eqn{s = 100 \cdot (y - \min(y)) /
#' (\max(y) - \min(y))}, using the min and max of the point estimates.
#' CI bounds are clamped to \eqn{[0, 100]}.
#'
#' @param f_hat_star Length-N numeric vector of corrected composite scores.
#' @param ci_lower Length-N numeric vector of 2.5th-percentile bootstrap
#'   scores.
#' @param ci_upper Length-N numeric vector of 97.5th-percentile bootstrap
#'   scores.
#' @param lambda Numeric. Box-Cox power parameter (default 0.5).
#' @param eps Numeric. Small constant to ensure positivity before Box-Cox
#'   (default 1e-6).
#'
#' @return A named list with:
#'   \itemize{
#'     \item `score`: length-N rescaled point estimates in \eqn{[0, 100]}.
#'     \item `ci_lower`: length-N rescaled lower CI bounds (clamped to
#'       \eqn{[0, 100]}).
#'     \item `ci_upper`: length-N rescaled upper CI bounds (clamped to
#'       \eqn{[0, 100]}).
#'   }
#' @export
postprocess_scores <- function(f_hat_star, ci_lower, ci_upper,
                                lambda = 0.5, eps = 1e-6) {
  # ---- Step 1: Z-score ----
  mu  <- mean(f_hat_star, na.rm = TRUE)
  sig <- stats::sd(f_hat_star)

  z_score <- (f_hat_star - mu) / sig
  z_lower <- (ci_lower   - mu) / sig
  z_upper <- (ci_upper   - mu) / sig

  # ---- Step 2: Box-Cox (lambda = 0.5) ----
  # Shift to strictly positive domain using min of point estimates
  shift   <- abs(min(z_score, na.rm = TRUE)) + eps
  x_score <- z_score + shift
  x_lower <- z_lower + shift
  x_upper <- z_upper + shift

  box_cox <- function(x) (pmax(x, 0) ^ lambda - 1) / lambda

  bc_score <- box_cox(x_score)
  bc_lower <- box_cox(x_lower)
  bc_upper <- box_cox(x_upper)

  # ---- Step 3: Min-max rescale to [0, 100] ----
  bc_min <- min(bc_score, na.rm = TRUE)
  bc_max <- max(bc_score, na.rm = TRUE)
  rng    <- bc_max - bc_min

  rescale <- function(x) 100 * (x - bc_min) / rng

  score    <- rescale(bc_score)
  ci_lo    <- pmin(pmax(rescale(bc_lower), 0), 100)
  ci_up    <- pmin(pmax(rescale(bc_upper), 0), 100)

  list(score = score, ci_lower = ci_lo, ci_upper = ci_up)
}

File: python/mpca/postprocess.py
"""Post-processing: z-score → Box-Cox → min-max rescaling."""

import numpy as np


def postprocess_scores(
    f_hat_star: np.ndarray,
    ci_lower: np.ndarray,
    ci_upper: np.ndarray,
    lam: float = 0.5,
    eps: float = 1e-6,
) -> dict:
    """
    Apply three-step post-processing to composite scores and CI bounds.

    **Step 1 — Z-score**: subtract the mean and divide by the SD of the
    point estimates.  The same shift and scale are applied to CI bounds.

    **Step 2 — Box-Cox** (:math:`\\lambda = 0.5`): shift all values to a
    strictly positive domain (:math:`x \\leftarrow x + |\\min(x)| + \\epsilon`
    using the minimum of the point estimates), then apply
    :math:`y = (x^\\lambda - 1)/\\lambda`.

    **Step 3 — Min-max** to :math:`[0, 100]`: use the min and max of the
    transformed *point estimates*; CI bounds are clamped to
    :math:`[0, 100]`.

    Parameters
    ----------
    f_hat_star : np.ndarray, shape (N,)
        Corrected composite point estimates.
    ci_lower : np.ndarray, shape (N,)
        2.5th-percentile bootstrap scores.
    ci_upper : np.ndarray, shape (N,)
        97.5th-percentile bootstrap scores.
    lam : float, optional
        Box-Cox power parameter (default 0.5).
    eps : float, optional
        Small constant for positivity before Box-Cox (default 1e-6).

    Returns
    -------
    dict with keys:

    ``score`` : np.ndarray, shape (N,)
        Rescaled point estimates in :math:`[0, 100]`.
    ``ci_lower`` : np.ndarray, shape (N,)
        Rescaled lower CI bounds, clamped to :math:`[0, 100]`.
    ``ci_upper`` : np.ndarray, shape (N,)
        Rescaled upper CI bounds, clamped to :math:`[0, 100]`.
    """
    # ---- Step 1: Z-score ----
    mu = np.nanmean(f_hat_star)
    sig = np.nanstd(f_hat_star, ddof=1)

    z_score = (f_hat_star - mu) / sig
    z_lower = (ci_lower - mu) / sig
    z_upper = (ci_upper - mu) / sig

    # ---- Step 2: Box-Cox ----
    shift = abs(np.nanmin(z_score)) + eps
    x_score = z_score + shift
    x_lower = z_lower + shift
    x_upper = z_upper + shift

    def box_cox(x: np.ndarray) -> np.ndarray:
        return (np.maximum(x, 0) ** lam - 1) / lam

    bc_score = box_cox(x_score)
    bc_lower = box_cox(x_lower)
    bc_upper = box_cox(x_upper)

    # ---- Step 3: Min-max to [0, 100] ----
    bc_min = np.nanmin(bc_score)
    bc_max = np.nanmax(bc_score)
    rng = bc_max - bc_min

    def rescale(x: np.ndarray) -> np.ndarray:
        return 100.0 * (x - bc_min) / rng

    score = rescale(bc_score)
    ci_lo = np.clip(rescale(bc_lower), 0.0, 100.0)
    ci_up = np.clip(rescale(bc_upper), 0.0, 100.0)

    return {"score": score, "ci_lower": ci_lo, "ci_upper": ci_up}
